lib/collection/src/grouping/aggregator.rs

```diff
@@ -0,0 +1,425 @@
+    fn add_point(&mut self, point: &ScoredPoint) -> Result<(), AggregatorError> {
+        if self.full_groups.len() == self.max_groups {
+            return Err(AllGroupsFull);
+        }
+
+        // if the key contains multiple values, grabs the first one
+        let group_key = point
+            .payload
+            .as_ref()
+            .and_then(|p| p.get_value(&self.grouped_by).into_iter().next().cloned())
+            .ok_or(KeyNotFound)
+            .and_then(GroupKey::try_from)?;
+
+        let group = self
+            .groups
+            .entry(group_key.clone())
+            .or_insert_with(|| HashSet::with_capacity(self.max_group_size));
+
+        group.insert(HashablePoint::minimal_from(point));
+
+        if group.len() == self.max_group_size {
+            self.full_groups.insert(group_key);
+        }
+
+        Ok(())
+    }
+
+    /// Adds multiple points to the group that they corresponds based on the group_by field, assumes that the points always have the grouped_by field, else it just ignores them
+    pub(super) fn add_points(&mut self, points: &[ScoredPoint]) {
+        for point in points {
+            if let Err(AllGroupsFull) = self.add_point(point) {
+                break;
+            }
+        }
+    }
+
+    #[cfg(test)]
+    pub(super) fn len(&self) -> usize {
+        self.groups.len()
+    }
+
+    // Gets the keys of the groups that have less than the max group size
+    pub(super) fn keys_of_unfilled_groups(&self) -> Vec<Value> {
+        self.groups
+            .iter()
+            .filter(|(_, hits)| hits.len() < self.max_group_size)
+            .map(|(key, _)| Value::from(key.clone()))
+            .collect()
+    }
+
+    // gets the keys of the groups that have reached the max group size
+    pub(super) fn keys_of_filled_groups(&self) -> Vec<Value> {
+        self.full_groups.iter().cloned().map(Value::from).collect()
+    }
+
+    /// Gets the ids of the already present points across all the groups
+    pub(super) fn ids(&self) -> HashSet<ExtendedPointId> {
+        self.groups
+            .iter()
+            .flat_map(|(_, hits)| hits.iter())
+            .map(|p| p.id())
+            .collect()
+    }
+
+    // Returns the best groups sorted by their best hit. The hits are sorted too
+    pub(super) fn distill(&self) -> Vec<Group> {
+        self.groups
+            .iter()
+            .map(|(k, v)| {
+                let hits = v
+                    .iter()
+                    .sorted()
+                    .rev()
+                    .take(self.max_group_size)
+                    .cloned()
+                    .collect::<Vec<_>>();
+
+                Group {
+                    hits,
+                    key: k.clone(),
+                    group_by: self.grouped_by.clone(),
+                }
+            })
+            // TODO: what happens when the best groups are not the most filled ones?
```

Comment:
I believe we should select groups to returned according to their max scores and not according to their size.

Reply:
Fixed this using a priority queue and modifying the `group_by` method accordingly